src/Ledger/Epoch.agda

```diff
@@ -6,99 +6,91 @@ open import Ledger.Prelude hiding (compare; Rel)
+  <ᵉ-resp-≡ : _<ᵉ_ Respects₂ _≡_
+  <ᵉ-resp-≡ = i , ii
+    where
+    i : _<ᵉ_ Respectsʳ _≡_
+    i refl x<y = x<y
+    ii : _<ᵉ_ Respectsˡ _≡_
+    ii refl y<x = y<x
 
   instance
-    preoSlot : HasPreorder Slot _≡_
-    preoSlot = record { _≤_ = _≤ˢ_ ; isPreorder = ≤ˢ-isPreorder }
-
-    poSlot : HasPartialOrder Slot _≡_
-    poSlot = record { hasPreorder = preoSlot ; antisym = ≤ˢ-isAntisymmetric }
+    preoEpoch : HasPreorder≡ Epoch
+    preoEpoch = hasPreorderFromStrictPreorder Epoch _≡_ _<ᵉ_ isEquivalence <ᵉ-irrefl <-trans <ᵉ-resp-≡
 
-  Dec≡⋀TotAntisym≤⇒Dec≤ : Decidable _≡_ → Antisymmetric _≡_ _≤ˢ_ → Total _≤ˢ_ → Decidable _≤ˢ_
-  Dec≡⋀TotAntisym≤⇒Dec≤ dec≡ antisym≤ tot≤ x y with dec≡ x y | tot≤ x y
-  ... | yes refl | _ = true because ofʸ (IsPartialOrder.reflexive ≤ˢ-isPartialOrder refl)
-  ... | no ¬p | inj₁ x≤y = true because ofʸ x≤y
-  ... | no ¬p | inj₂ y≤x = false because (ofⁿ (λ x≤y → ¬p (antisym≤ x≤y y≤x)))
+  ≤ˢ-transitive : Transitive _≤_
+  ≤ˢ-transitive = IsPreorder.trans ≤-isPreorder
 
+  ≤ˢ-isTotalOrder : IsTotalOrder _≡_ _≤_
+  ≤ˢ-isTotalOrder = <-STO⇒≤-isTotalOrder Slot _≡_ _<_ Slot-STO
 
-  _≤ˢ?_ : (s s' : Slot) → Dec (s ≤ˢ s')
-  _≤ˢ?_ = Dec≡⋀TotAntisym≤⇒Dec≤ _≡ˢ?_ ≤ˢ-isAntisymmetric ≤ˢ-isTotal
+  ≤ˢ-total : Total _≤_
+  ≤ˢ-total = IsTotalOrder.total ≤ˢ-isTotalOrder
 
-  ≤ˢ-isDecTotalOrder : IsDecTotalOrder _≡_ _≤ˢ_
-  ≤ˢ-isDecTotalOrder = record { isTotalOrder = Slot-TO ; _≟_ = _≡ˢ?_ ; _≤?_ = _≤ˢ?_ }
-
-  instance
-    Dec-≤ˢ : ∀ {n m} → Dec (n ≤ˢ m)
-    Dec-≤ˢ = Decidable²⇒Dec _≤ˢ?_
-
-    decpoSlot : HasDecPartialOrder Slot _≡_
-    decpoSlot = record { hasPartialOrder = poSlot ; _≤?_ = _≤ˢ?_ }
+  ≤ˢ-isDecTotalOrder : IsDecTotalOrder _≡_ _≤_
+  ≤ˢ-isDecTotalOrder = record { isTotalOrder = ≤ˢ-isTotalOrder ; _≟_ = _≟_ ; _≤?_ = _≤?_ }
 
   _≤ᵉ_ : Epoch → Epoch → Set
-  e ≤ᵉ e' = firstSlot e ≤ˢ firstSlot e'
-
-  ≤ᵉ-isReflexive : Reflexive _≤ᵉ_
-  ≤ᵉ-isReflexive = IsPreorder.reflexive ≤ˢ-isPreorder refl
+  e ≤ᵉ e' = firstSlot e ≤ firstSlot e'
 
-  _≤ᵉ?_ : ∀ e e' → Dec (e ≤ᵉ e')
-  e ≤ᵉ? e' = firstSlot e ≤ˢ? firstSlot e'
+  ≤ᵉ-reflexive : Reflexive _≤ᵉ_
+  ≤ᵉ-reflexive = IsPreorder.reflexive ≤-isPreorder refl
 
   ≤ᵉ-isPreorder : IsPreorder _≡_ _≤ᵉ_
-  ≤ᵉ-isPreorder .IsPreorder.isEquivalence     = Ledger.Prelude.isEquivalence
-  ≤ᵉ-isPreorder .IsPreorder.reflexive refl    = ≤ᵉ-isReflexive
-  ≤ᵉ-isPreorder .IsPreorder.trans ij jk       = ≤ˢ-isTransitive ij jk
+  ≤ᵉ-isPreorder .IsPreorder.isEquivalence   = Ledger.Prelude.isEquivalence
+  ≤ᵉ-isPreorder .IsPreorder.reflexive refl  = ≤ᵉ-reflexive
+  ≤ᵉ-isPreorder .IsPreorder.trans ij jk     = ≤ˢ-transitive ij jk
+
+  _≤ᵉ?_ : (e e' : Epoch) → Dec (e ≤ᵉ e')
+  e ≤ᵉ? e' = firstSlot e ≤? firstSlot e'
 
-  _ = (∀ {n m} → Dec (n <ˢ m)) ∋ it
-  _ = (∀ {n m} → Dec (n ≤ˢ m)) ∋ it
-  _ = (∀ {n m} → Dec (n ≤ᵉ m)) ∋ it
 
   instance
-    preoEpoch : HasPreorder Epoch _≡_
-    preoEpoch = record { _≤_ = _≤ᵉ_ ; isPreorder = ≤ᵉ-isPreorder }
+    Dec-<ˢ : ∀ {n m : Slot} → Dec (n <ˢ m)
```

Comment:
This should be using `HasDecPartialOrder≡`. I think you've previously mention that this doesn't work since it's not antisymmetric, but it is in all the cases we care about. So you can just add a field `firstSlot-injective`, and then you can prove that the order on epochs is antisymmetric.